modules/heat_conduction/src/materials/SemiconductorLinearConductivity.C

```diff
@@ -0,0 +1,54 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "SemiconductorLinearConductivity.h"
+#include "libmesh/quadrature.h"
+
+template <>
+InputParameters
+validParams<SemiconductorLinearConductivity>()
+{
+  InputParameters params = validParams<Material>();
+  params.addRequiredCoupledVar("temp", "Variable for temperature.");
+  params.addParam<std::string>("base_name", "Material property base name");
+  params.addRequiredParam<Real>("sh_coeff_A", "Steinhart_Hart coefficient A of the material");
+  params.addRequiredRangeCheckedParam<Real>(
+      "sh_coeff_B",
+      "sh_coeff_B != 0",
+      "Steinhart_Hart coefficient B of the material.  Tt should "
+      "not be zero as it is in denominator.");
+  return params;
+}
+
+SemiconductorLinearConductivity::SemiconductorLinearConductivity(const InputParameters & parameters)
+  : DerivativeMaterialInterface<Material>(parameters),
+    _sh_coeff_A(getParam<Real>("sh_coeff_A")),
+    _sh_coeff_B(getParam<Real>("sh_coeff_B")),
+    _T(coupledValue("temp")),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _electric_conductivity(declareProperty<Real>(_base_name + "electrical_conductivity")),
+    _delectric_conductivity_dT(declarePropertyDerivative<Real>(
+        _base_name + "electrical_conductivity", getVar("temp", 0)->name()))
+{
+}
+
+void
+SemiconductorLinearConductivity::computeQpProperties()
+{
+  mooseAssert(_T[_qp] > 0, "Negative value for temperature in Kelvin or divided by zero");
```

Comment:
As I was saying before, an assert on _T is not sufficient here because asserts are only active in debug mode, but it would be really easy for a user to set up a model that would violate this assert, and there would be no check for this in an optimized executable.  Replace the above line with this:
```
if (MooseUtils::absoluteFuzzyLessEqual(_T[_qp], 0.0))
  mooseError("Encountered zero or negative temperature in SemiconductorLinearConductivity");
```

Reply:
This debate will never be settled, but I'll add my two cents:

Adding a branch and an fuzzy check is likely to cause a fairly significant slowdown since this is a per quadrature point check. You are absolutely correct that the assert is not sufficient since an optimized model would miss it and trip an FPE. However, in practice, we'd catch that kind of thing with a simple debug or even devel check (which is highly encouraged for "DEVELopers" anyway), which is the purpose of the assert. 

I'm not a fan of expensive runtime checks for things that shouldn't happen in a correctly setup model. Do we really end up with a zero temperature that often once a simulation is running? If not, then the full branch and check is probably overkill. I'm just trying to get people thinking about asserts a lot more often in their code. They are "free" in optimized mode so go nuts with them. Use your optimized mode checks sparingly for things that can happen during the simulation that might be beyond simple model setup problems.

Reply:
Thanks for the input. I guess it really comes down to just how expensive the check is, and how likely you are to encounter that corner case that you are checking for.

I'm never really sure how expensive these kinds of checks are to make a well-informed decision. We tend to err on the side of including them. It's not usually developers that hit these checks -- it's end users that are generally running the opt executable.  I guess my thinking is that if it's an error, the end user will recognize the problem and fix it without us even hearing about it, whereas if it's an assert, they'll probably get frustrated with it, send it to us, and then we'll run it in dbg and discover the problem, a process that could waste way more of everyone's time than the runtime overhead of the error check.

I'm really curious -- is this check really _that_ expensive? This seems like kind of a corner case, although we actually hit it on the first problem that was ever run with this model -- the regression test (because the temperature field was initially zero, which is the default). However, in that case, even though there was a divide by zero, the model happily computed the correct solution, which is why we didn't even notice we had a problem until the test was run with the debug executable (which actually exposed the error without us needing to add an assert because it generated a floating point exception).

This is a lot of discussion about one line of code, but it's an important issue that warrants it because we deal with this all over the place.

Anyway, @yeh512, weighing everything out, I think you should just leave that as an assert like you had it. I don't think we need to hold up your commit while we debate about what to do in general about this kind of thing. I still would like you to address my other comment, though.

Reply:
> I'm really curious -- is this check really that expensive?

I'm going to tag @rwcarlsen in here. While it's definitely good advice to not worry about optimization during development, its a good idea to learn how to start measuring these things. For that you can always load a profiling tool and just run your input files with and without these checks to directly measure. We recently merged in a new "benchmarks" capability where we run the simulation several times to produce a statistical plots. We can directly see these affects. Material properties are prime suspects for chewing up CPU time. They are called more often than any other object in MOOSE. Why? because they are evaluated on demand for each system that consumes them. This might happen multiple times for each residual evaluation, and there is at least one residual evaluation for each linear solve so yes, they "can" be that expense. Branching can stall the processor and cost you more than you might expect. Multiply that by several million (num evals * element * qps * linear its) and it adds up.

Reply:
Or how about:

```c++
mooseAssert(MooseUtils::absoluteFuzzyLessEqual(_T[_qp], 0.0), "Negative value for temperature in Kelvin or divided by zero"
```

I agree with cody that we probably don't want this sort of runtime check here in a ~hotter area of moose.  The ``[moose]/scripts/README.md`` file has some directions on hacking up and running some quick manual/custom benchmarks.  But you will need to do it on a very quiet machine for good results.